viewport/gui_in_3d/gui_3d.gd

```diff
@@ -1,102 +1,156 @@
-	event.global_position = pos
+	event.position = mouse_pos2D
+	event.global_position = mouse_pos2D
 	
 	# If the event is a mouse motion event...
 	if event is InputEventMouseMotion:
 		# If there is not a stored previous position, then we'll assume there is no relative motion.
-		if prev_pos == null:
+		if last_mouse_pos2D == null:
 			event.relative = Vector2(0, 0)
 		# If there is a stored previous position, then we'll calculate the relative position by subtracting
 		# the previous position from the new position. This will give us the distance the event traveled from prev_pos
 		else:
-			event.relative = pos - prev_pos
-	
-	# Update prev_pos with the position we just calculated.
-	prev_pos = pos
+			event.relative = mouse_pos2D - last_mouse_pos2D
+	# Update last_mouse_pos2D with the position we just calculated.
+	last_mouse_pos2D = mouse_pos2D
 	
 	# Finally, send the processed input event to the viewport.
-	viewport.input(event)
+	node_viewport.input(event)
 
+func find_mouse(global_position):
+	var camera = get_viewport().get_camera()
+	
+	#from camera center to the mouse position in the Area
+	var from = camera.project_ray_origin(global_position)
+	var dist = find_further_distance_to(camera.transform.origin)
+	var to = from + camera.project_ray_normal(global_position) * dist
+	
+	
+	#Manually raycasts the are to find the mouse position
+	var result = get_world().direct_space_state.intersect_ray(from, to, [], node_area.collision_layer) #,false,true) #for 3.1 changes
+	
+	if result.size() > 0:
+		return result.position
+	else:
+		return null
 
-func _ready():
-	# Get the Viewport node and assign it to viewport for later use.
-	viewport = get_node("Viewport")
-	# Connect the input_event signal to the _on_area_input_event function.
-	get_node("Area").connect("input_event", self, "_on_area_input_event")
-  
+func find_further_distance_to(origin):
+	#Find edges of collision and change to global positions
+	var edges = []
+	edges.append(node_area.to_global(Vector3(quad_mesh_size.x / 2, quad_mesh_size.y / 2, 0)))
+	edges.append(node_area.to_global(Vector3(quad_mesh_size.x / 2, -quad_mesh_size.y / 2, 0)))
+	edges.append(node_area.to_global(Vector3(-quad_mesh_size.x / 2, quad_mesh_size.y / 2, 0)))
+	edges.append(node_area.to_global(Vector3(-quad_mesh_size.x / 2, -quad_mesh_size.y / 2, 0)))
+	
+	#Get the furthest distance between the camera and collision to avoid raycasting too far or too short
+	var far_dist = 0
+	var temp_dist
+	for edge in edges:
+		temp_dist = origin.distance_to(edge)
+		if temp_dist > far_dist:
+			far_dist = temp_dist
+	
+	return far_dist
+
+func rotate_area_to_billboard():
+	var billboard_mode = node_quad.get_surface_material(0).params_billboard_mode
+	
+	#try to match the area with the material's billboard setting, if enabled
+	if billboard_mode > 0:
+		#Look in the same direction as the camera
+		var look = get_viewport().get_camera().to_global(Vector3(0,0,-100)) - get_viewport().get_camera().global_transform.origin
```

Comment:
I'm sure this works as intended, but I don't quite understand what this line is doing or what it is for exactly. Why are you getting a position 100 units in front of the camera relative to the camera, but then subtracting the camera's position?

If I recall correctly, doesn't `look_at` take a position in global/world space? Shouldn't `get_viewport().get_camera().global_transform.origin` work?

I'm just curious because I cannot quite figure out what this line of code and the one below it are doing.

Reply:
The billboard setting don't look at the camera, instead it looks at the inverse direction the camera is looking.

I used -100 because `look_at` was complaining about "being in the same position as the object". So I threw a big number to make sure it wasn't going to complain.

What this code does:
1. Get a position in front of the camera
2. Change it to global to find the direction it's looking at.
3. Remove the camera offset to get a "clean" looking direction coming from the world origin.
4. Finally, adds the area offset to make it look at the exact inverted direction* the camera is facing.
5. In case of Y-Billboard, lock Y axis to 0.

*Since Godot uses forward as -1 and the area is already facing the other way, I didn't need to invert anything.

There might be a simpler way to do this with more knowledge in vector math, but this is the best I could find.

Reply:
Ah, okay. That makes sense. Thanks for explaining! Not my lack of understanding would have changed anything, but I think I now I more or less know what the code is doing now. Thanks again for explaining!